text/0000-min-rust-version.md

```diff
@@ -0,0 +1,205 @@
+extension)
+- version should not be bigger than the current stable toolchain
+- version should not be smaller than 1.27 (version in which  `package.rust` field
+became a warning instead of an error)
+
+`rust` will be a required field. For crates uploaded before introduction of this
+feature 2015 edition crates will imply `rust="1.0"` and 2018 edition will imply
+`rust = "1.30"`.
+
+It will be an error to use `rust="1.27"` and `edition="2018"`, but `rust="1.40"`
+and `edition="2015"` is a valid combination.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The describe functionality can be introduced in several stages:
+
+
+## First stage: dumb field
+
+At first the `rust` field can be simply a declarative optional field without any
+functionality behind it with minimal checks. The reason for it is to reduce
+implementation cost of the first stage to the minimum and ideally ship it as part
+of Rust 2018. It will also allow crate authors who care about MSRV to start mark
+their crates early.
+
+## Second stage: `cargo publish` check
+
+The next step is for `cargo publish` to require use of the toolchain specified
+in the `rust` field, for example crates with:
+- `rust="stable"` can be published only with a stable toolchain, though not
+necessarily with the latest one. Cargo will insert toolchain version before
+publishing the crate as was described in the "guide-level explanation".
+- `rust="nightly: *"` can be published only with a nightly toolchain. If finer
+grained "nightly: ..." (see "nightly versions" section) is selected, then one
+of the selected Nightly versions will have to be used.
+- `rust="1.30"` can be published only with (stable) Rust 1.30, even if it's
+not the latest stable Rust version.
+
+Using the usual build check `cargo publish` will verify that crate indeed can be
+built using specified MSRV. This check can be used with exisiting `--no-verify`
+option.
+
+## Third stage: versions resolution
+
+`rust` field will be used as a constraint for dependency versions resolution.
+If user uses e.g. Rust 1.40 and uses crate `foo = "0.2"`, but
+all selected versions of `foo` specify MSRV e.g. equal 1.41 or bigger (or even
+nightly) `cargo` will issue an error.
+
+`rust` field value will be checked as well, on crate build `cargo` will check if
+all upstream dependencies can be built with the specified MSRV. (i.e. it will
+check if there is exists solution for given crates and Rust versions constraints)
+
+Yanked crates will be ignored in this process.
+
+Implementing this functionality hopefully will allow to close the debate regarding
+MSRV handling in crate versions and will allow crate authors to feel less
+restrictive about bumping their crate's MSRV. (though it can be a usefull
+convention for post-1.0 crates to bump minor version on MSRV change to allow
+publishing backports which fix serious issues using patch version)
+
+## Extension: nightly versions
```

Comment:
I like the idea of flagging supported compiler versions, but I don't quite understand how a crate would actually be able to use this in practice if the list of supported versions has to go in the published Cargo.toml file itself.

Typically when publishing a nightly only crate, I know it works for the current nightly and that it will continue to work for all future nightlies until there is a breaking change. With this design, it seems like I'd need to publish a new version of the crate every day with an updated Cargo.toml saying that the new nightly also works or else my users would be unable to upgrade their compiler.

Reply:
As I see it, the main use-case will be to provide exact nightly version, on which crate is developed and tested. So e.g. if you depend on `rocket x.y.z` which is developed and published with nightly-2019-02-01 toolchain, then your crate will have to be developed with the same toolchain. This will result in a soft synchronization of nightly versions used across nightly crates ecosystem. (e.g. crates will update nightlies ever 2-3 weeks or so, not every night)

If your crate does not experience breakage too often, then you can use default `nightly: *` option.

Reply:
Using the field to indicate a known good nightly version seems reasonable. However, I'd be a concerned about using it for version resolution. The entire ecosystem of nightly crates would have to update in lockstep or else become mutually incompatible. Further, trying out new nightlies would require waiting for all your dependencies to be updated. Basically, you'd no longer be running on the nightly channel but instead a sort of "monthly" channel which lacked both the stability of stable and the quick bug fixes of nightly.

Reply:
I think most of the nightly crates will just use "nightly: *", some may use ranges (though properly testing it will not be easy), finer grained nightly selection will be used only be a handful of crates with rare intersections with each other. Plus do not forget that you always can disable MSRV constraint in dependency versions resolution with `--no-msrv-check` flag.

Reply:
I'm sorry, I'm still not fully understanding. Lets use a concrete example. I maintain the [`rahashmap`](crates.io/crates/rahashmap) crate which depends on some rather unstable features and so breaks every couple weeks. The last time this happened, I fixed the issue and promptly published a new version. This new version supports nightly-2018-07-10, nightly-2018-07-11, and all future nightlies until another breaking change happens. If this RFC & extension were implemented, what would the correct value for MSRV?

Reply:
The problem with "all future nightlies until another breaking change happens" is that this constraint can be properly expressed at the moment of crate publishing, as you don't know when breaking change will happen. The closest thing which we can get is to use `rust="nightly: >= 2018-07-10"` (it's mentioned in the extension). On the next breaking change you'll publish `rust="nightly: >= 2018-08-10"`. So assuming that user keeps all his dependencies updated, with new toolchain he'll use newer crate version, and on old toolchain the older crate version will be used. But `nightly: *` will not be much worse and will require significantly less effort from you.

But this approach feels quite fragile. To reliably solve this problem we will need an additional channel for notifying users "this crate broke on the following nightly versions" which can't be done via `Cargo.toml`.